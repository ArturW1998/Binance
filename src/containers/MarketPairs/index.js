import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import cn from 'classnames';
import { PATHS, LISTS } from '../../constants';
import {
  setActiveMarket,
  toggleSocketStreams,
  updateMarketPairs,
} from '../../actions';
import { getTickerBySymbol } from '../../utils';
import Loader from '../../components/Loader';
import Table from '../../components/Table';

const { PAIRS_LIST } = LISTS;

export class MarketPairs extends Component {
  static propTypes = {
    marketPairs: PropTypes.objectOf(PropTypes.objectOf(PropTypes.string)),
    activeMarket: PropTypes.shape({
      market: PropTypes.string,
    }),
    connectSocket: PropTypes.bool,
    setActiveMarket: PropTypes.func,
    updateMarketPairs: PropTypes.func,
    toggleSocketStreams: PropTypes.func,
  };

  static defaultProps = {
    marketPairs: {},
    activeMarket: {},
    connectSocket: true,
    setActiveMarket: () => {},
    updateMarketPairs: () => {},
    toggleSocketStreams: () => {},
  };

  constructor(props) {
    super(props);

    const {
      marketPairs,
      activeMarket: { market },
    } = props;

    this.state = {
      isLoaded: marketPairs && market,
    };

    this.streams = ['!miniTicker@arr'];
  }

  componentDidMount() {
    const { connectSocket } = this.props;

    connectSocket && this.connectSocketStreams(this.streams);
  }

  componentWillUnmount() {
    const { connectSocket } = this.props;

    connectSocket && this.disconnectSocketStreams(this.streams);
  }

  switchSocketStreams = () => {
    const { connectSocket } = this.props;

    connectSocket
      ? this.disconnectSocketStreams(this.streams)
      : this.connectSocketStreams(this.streams);
  };

  connectSocketStreams = streams => {
    const { BASE_PATH, STREAM_PATH, STREAM_PARAM } = PATHS;
    const joinedStreams = streams.join('/');
    const connection = btoa(joinedStreams);

    this[connection] = new WebSocket(
      `${BASE_PATH}${STREAM_PATH}?${STREAM_PARAM}${joinedStreams}`,
    );

    this[connection].onopen = () => {
      const { toggleSocketStreams } = this.props;

      toggleSocketStreams(true);
    };

    this[connection].onmessage = ({ data = {} }) => {
      const {
        updateMarketPairs,
        activeMarket: { market },
      } = this.props;
      const ticker = getTickerBySymbol(JSON.parse(data).data) || {};
      const BTC = PAIRS_LIST[1];

      updateMarketPairs(ticker);

      !market && this.setActiveTab(BTC);

      this.setState({
        isLoaded: true,
      });
    };

    this[connection].onerror = evt => {
      console.error(evt); // eslint-disable-line
    };
  };

  disconnectSocketStreams = streams => {
    const joinedStreams = streams.join('/');
    const connection = btoa(joinedStreams);

    this[connection].readyState === WebSocket.OPEN && this[connection].close();

    this[connection].onclose = () => {
      const { toggleSocketStreams } = this.props;

      toggleSocketStreams(false);
    };
  };

  setActiveTab = e => {
    const { setActiveMarket } = this.props;

    const market = e.currentTarget ? e.currentTarget.dataset.tab : e;

    const data = {
      market,
    };

    setActiveMarket(data);
  };

  render() {
    const { error, isLoaded } = this.state;
    const {
      activeMarket: { market },
      marketPairs,
      connectSocket,
    } = this.props;

    if (error) {
      return <div className="alert alert-danger">{error.message}</div>;
    }

    if (!isLoaded) {
      return <Loader />;
    }

    return (
      <>
        <h2 className="text-center pt-2">
          The World&apos;s Leading Cryptocurrency Exchange
        </h2>

        <div className="d-flex justify-content-center pt-2">
          <button
            className="btn btn-warning"
            onClick={this.switchSocketStreams}
            data-testid="Connect"
          >
            {connectSocket ? 'Disconnect' : 'Connect'}
          </button>
        </div>

        <ul className="nav nav-tabs pt-2" data-testid="NavGrid">
          {PAIRS_LIST.map(pair => {
            const classes = cn('nav-link', { active: market === pair });

            return (
              <li key={pair} className="nav-item">
                <button
                  className={classes}
                  onClick={this.setActiveTab}
                  data-tab={pair}
                >
                  {pair}
                  <span className="d-none d-sm-inline"> Markets</span>
                </button>
              </li>
            );
          })}
        </ul>

        {marketPairs && market && (
          <Table ticker={marketPairs} filter={market} />
        )}
      </>
    );
  }
}

export default connect(
  ({ marketPairs, activeMarket, connectSocket }) => ({
    marketPairs,
    activeMarket,
    connectSocket,
  }),
  {
    setActiveMarket,
    updateMarketPairs,
    toggleSocketStreams,
  },
)(MarketPairs);
