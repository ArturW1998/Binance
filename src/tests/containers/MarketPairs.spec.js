import React from 'react';
import { shallow } from 'enzyme';
import { MarketPairs } from '../../containers/MarketPairs';

describe('MarketPairs container', () => {
  const props = {
    marketPairs: {},
    activeMarket: {},
    connectSocket: false,
    setActiveMarket: () => {},
    updateMarketPairs: () => {},
    toggleSocketStreams: () => {},
  };

  describe('MarketPairs container initial', () => {
    const marketPairs = shallow(<MarketPairs {...props} />);

    it('renders initial', () => {
      expect(marketPairs.find('Loader')).toHaveLength(1);
    });

    it('not render <h2>', () => {
      expect(marketPairs.find('h2')).toHaveLength(0);
    });

    it('not render <Table />', () => {
      expect(marketPairs.find('Table')).toHaveLength(0);
    });

    it('renders properly', () => {
      expect(marketPairs).toMatchSnapshot();
    });
  });

  describe('MarketPairs container state', () => {
    const {
      marketPairs: marketPairsProps,
      activeMarket: { market: marketProps },
    } = props;

    const initialState = {
      isLoaded: marketPairsProps && marketProps,
      error: undefined,
    };

    const marketPairs = shallow(<MarketPairs {...props} />);

    it('initialize MarketPairs with initial state', () => {
      marketPairs.setState(initialState);

      expect(marketPairs.state()).toEqual(initialState);
    });

    describe('State handlers', () => {
      describe('MarketPairs container with error', () => {
        const nextState = {
          ...initialState,
          error: {
            message: 'Request failed with status code 400',
          },
        };

        it('initialize MarketPairs with error', () => {
          marketPairs.setState(nextState);

          expect(marketPairs.state().error).toEqual(nextState.error);
        });

        it('renders error', () => {
          expect(marketPairs.find('.alert').text()).toEqual(
            'Request failed with status code 400',
          );
        });

        it('render only one <div>', () => {
          expect(marketPairs.find('div')).toHaveLength(1);
        });

        it('not render <Table />', () => {
          expect(marketPairs.find('Table')).toHaveLength(0);
        });
      });

      describe('MarketPairs container isLoaded', () => {
        it('initialize MarketPairs with isLoaded', () => {
          const nextState = {
            ...initialState,
            isLoaded: true,
          };

          marketPairs.setState(nextState);

          expect(marketPairs.state().isLoaded).toEqual(nextState.isLoaded);
        });

        it('renders <Loader />', () => {
          const nextState = {
            ...initialState,
            isLoaded: true,
          };

          marketPairs.setState(nextState);

          expect(marketPairs.find('Loader')).toHaveLength(0);
        });

        it('initialize MarketPairs without isLoaded', () => {
          const nextState = {
            ...initialState,
            isLoaded: false,
          };

          marketPairs.setState(nextState);

          expect(marketPairs.find('Loader')).toHaveLength(1);
        });

        afterEach(() => {
          marketPairs.setState(initialState);
        });
      });

      afterAll(() => {
        marketPairs.setState(initialState);
      });
    });
  });

  describe('MarketPairs container setActiveMarket', () => {
    const mockSetActiveMarket = jest.fn();
    const nextProps = {
      ...props,
      activeMarket: {
        market: 'BTC',
      },
      marketPairs: {
        BNBBTC: {},
        ETHUSDT: {},
        ENJBTC: {},
      },
      setActiveMarket: mockSetActiveMarket,
    };

    const marketPairs = shallow(<MarketPairs {...nextProps} />);

    describe('when clicking into tab button', () => {
      const tab = 'BNB';

      beforeEach(() => {
        marketPairs.find(`[data-tab="${tab}"]`).simulate('click', {
          currentTarget: {
            dataset: {
              tab,
            },
          },
        });
      });

      it('dispatches the `setActiveMarket()` method it receives from props', () => {
        expect(mockSetActiveMarket).toHaveBeenCalledTimes(1);
      });
    });

    describe('when setting active tab', () => {
      const tab = 'BNB';

      beforeEach(() => {
        marketPairs.find(`[data-tab="${tab}"]`).simulate('click', tab);
      });

      it('dispatches the `setActiveMarket()` method it receives from props', () => {
        expect(mockSetActiveMarket).toHaveBeenCalled();
      });
    });
  });

  describe('MarketPairs container is loaded', () => {
    const nextProps = {
      ...props,
      activeMarket: {
        market: 'BTC',
      },
      marketPairs: {
        BNBBTC: {},
        ETHUSDT: {},
        ENJBTC: {},
      },
    };

    const marketPairs = shallow(<MarketPairs {...nextProps} />);

    it('renders <Table /> template', () => {
      expect(marketPairs.find('Table')).toHaveLength(1);
    });

    it('not render <Loader />', () => {
      expect(marketPairs.find('Loader')).toHaveLength(0);
    });

    it('renders properly', () => {
      expect(marketPairs).toMatchSnapshot();
    });

    it('renders class active', () => {
      const { market } = nextProps.activeMarket;

      expect(
        marketPairs.find(`[data-tab="${market}"]`).prop('className'),
      ).toEqual('nav-link active');
    });
  });
});
