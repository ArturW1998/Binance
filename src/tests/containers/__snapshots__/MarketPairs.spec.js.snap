// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MarketPairs container Connected MarketPairs container renders properly 1`] = `ShallowWrapper {}`;

exports[`MarketPairs container MarketPairs container initial renders properly 1`] = `ShallowWrapper {}`;

exports[`MarketPairs container MarketPairs container is loaded renders properly 1`] = `ShallowWrapper {}`;
